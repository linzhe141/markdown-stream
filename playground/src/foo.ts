export const md = `好的！下面 Vue 3 的简单 \`v-for\` 示例。
这个示例会遍历一个数组，并将数组中的每一项渲染为一个列表项。
# 代码示例 *vue3 todo list*



### 代码说明
1. **\`v-for\` 指令**：\`v-for\` 用于遍历数组 \`items\`，并将每个元素渲染为一个 \`<li>\` 标签。
2. **\`key\` 属性**：\`key\` 是 Vue 用于跟踪每个节点的唯一标识，这里使用 \`index\` 作为 \`key\`。
3. **\`ref\`**：使用 \`ref\` 创建一个响应式数组 \`items\`。
4. **模板**：在模板中渲染一个标题和一个无序列表，列表项的内容是数组中的每个元素。

### 运行效果
页面会显示一个标题和一个列表，列表项为：
\`\`\`
1. 苹果
2. 香蕉
3. 橙子
4. 葡萄
5. 芒果
\`\`\`

你可以根据需要修改 \`items\` 数组的内容，列表会自动更新。希望这个示例对你有帮助！如果有其他问题，随时问我哦！`;
