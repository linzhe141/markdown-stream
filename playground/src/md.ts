export const md = `好的！下面 Vue 3 的简单 \`v-for\` 示例。这个示例会遍历一个数组，并将数组中的每一项渲染为一个列表项。

\`zzz\`

**bbb**

*aaa*

# 代码示例 *vue3 todo list*

\`\`\`html  filename="app.vue"
<div>aaaaaaa</div>
\`\`\`

### 代码说明
- **\`v-for\` 指令**：\`v-for\` 用于遍历数组 \`items\`，并将每个元素渲染为一个 \`<li>\` 标签。
- **\`key\` 属性**：\`key\` 是 Vue 用于跟踪每个节点的唯一标识，这里使用 \`index\` 作为 \`key\`。
- **\`ref\`**：使用 \`ref\` 创建一个响应式数组 \`items\`。
- **模板**：在模板中渲染一个标题和一个无序列表，列表项的内容是数组中的每个元素。

### 运行效果
页面会显示一个标题和一个列表，列表项为：
\`\`\`
1. 苹果
2. 香蕉
3. 橙子
4. 葡萄
5. 芒果
\`\`\`

你可以根据需要修改 \`items\` 数组的内容，列表会自动更新。希望这个示例对你有帮助！如果有其他问题，随时问我哦！`;
export const md1 = `
*foo\`bar\`zzzzzzzz*ggg

*xxxx*

*\`linzhe\`*

**fffff**

*sss*

\`zzz\`
`;
