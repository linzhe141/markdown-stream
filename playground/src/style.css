@import "tailwindcss";
@plugin "@tailwindcss/typography";
.dark .prose {
  --tw-prose-headings: white !important;
  --tw-prose-code: white !important;
  color: white;
}
