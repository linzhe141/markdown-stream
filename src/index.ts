import { ChunkData } from "./core/stream-tokenizer";
import { createStreamTransform } from "./core/stream-transform";

export async function createMarkdownStreamRender(
  stream: ReadableStream,
  container: Element
) {
  injectCopyToClipboard();
  const transformedStream = createStreamTransform(stream);

  const reader = transformedStream.getReader();
  const stack: Element[] = [];
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    const { type, chunk, operate } = value as ChunkData;
    if (operate === "blockOpen") {
      if (type === "text") {
        const dom = document.createElement("p");
        stack.push(dom);
        dom.textContent = value.chunk;
        container.appendChild(dom);
      } else if (type === "headingType") {
        const dom = document.createElement("h" + chunk.length);
        stack.push(dom);
        container.appendChild(dom);
      } else if (type === "inlineCodeOpen") {
        const dom = document.createElement("code");
        stack.push(dom);
        container.appendChild(dom);
      } else if (type === "codeBlockOpen") {
        const s = document.createElement("div");
        s.innerHTML = `<div class="code-block" style="border-color: #334155;border-radius: 4px;border: 1px solid;overflow: hidden;">
              <div style="padding:0 16px;display: flex;justify-content: space-between; background: #334155; position: sticky;top: 0;">
                <div class="lang code-block-meta"></div>
                <div onclick="_smd_copyToClipboard__(this)" style="display: flex; align-items: center;">
                  <svg width="12" height="12" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" class="icon-xs"><path d="M7 5C7 3.34315 8.34315 2 10 2H19C20.6569 2 22 3.34315 22 5V14C22 15.6569 20.6569 17 19 17H17V19C17 20.6569 15.6569 22 14 22H5C3.34315 22 2 20.6569 2 19V10C2 8.34315 3.34315 7 5 7H7V5ZM9 7H14C15.6569 7 17 8.34315 17 10V15H19C19.5523 15 20 14.5523 20 14V5C20 4.44772 19.5523 4 19 4H10C9.44772 4 9 4.44772 9 5V7ZM5 9C4.44772 9 4 9.44772 4 10V19C4 19.5523 4.44772 20 5 20H14C14.5523 20 15 19.5523 15 19V10C15 9.44772 14.5523 9 14 9H5Z" fill="currentColor"></path></svg>
                  <span style="margin-left: 8px;">copy</span>
                </div>
              </div>
              <pre style="margin:0"><code class="code-wrapper"></code></pre>
            </div>
          `;
        const dom = s.firstElementChild!;
        stack.push(dom);
        container.appendChild(dom);
      }
    } else {
      const last = stack[stack.length - 1];
      if (last) {
        if (type === "inlineCodeOpen") {
          const inlineCode = document.createElement("code");
          last.appendChild(inlineCode);
          stack.push(inlineCode);
        } else if (type === "inlineCode") {
          const inlineCode = stack[stack.length - 1];
          if (inlineCode) {
            const textNode = document.createTextNode(value.chunk);
            inlineCode.appendChild(textNode);
          }
        } else if (type === "inlineCodeClose") {
          stack.pop();
        } else if (type === "codeBlockMeta") {
          const codeBlockMeta = last.querySelector(".code-block-meta")!;
          codeBlockMeta.textContent = chunk;
        } else if (type === "codeBlock") {
          const codeWrapper = last.querySelector(".code-wrapper")!;
          const textNode = document.createTextNode(value.chunk);
          codeWrapper.appendChild(textNode);
        } else {
          const textNode = document.createTextNode(value.chunk);
          last.appendChild(textNode);
        }
      }
    }
  }
}

function injectCopyToClipboard() {
  if ((window as any)._smd_copyToClipboard__) return;
  (window as any)._smd_copyToClipboard__ = function (wrapper: Element) {
    const spanEl = wrapper.querySelector("span")!;
    spanEl.textContent = "copied!";
    setTimeout(() => (spanEl.textContent = "copy"), 2000);
    let nextSibling = wrapper.parentNode?.nextSibling;
    let preElement;
    while (nextSibling) {
      if (
        nextSibling.nodeType === 1 &&
        (nextSibling as Element).tagName === "PRE"
      ) {
        preElement = nextSibling;
        break;
      }
      nextSibling = nextSibling.nextSibling;
    }
    const textContent = preElement?.textContent;
    if (!textContent) return;
    navigator.clipboard.writeText(textContent);
  };
}
