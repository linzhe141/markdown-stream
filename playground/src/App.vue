<script setup lang="ts">
import { ref } from "vue";
import { md } from "./md";
import { createMarkdownStreamRender } from "../../src";
const container = ref<Element | null>(null);
function createStream(text: string, chunkSize = 10, delay = 50) {
  let position = 0;

  return new ReadableStream({
    pull(controller) {
      return new Promise((resolve) => {
        setTimeout(() => {
          if (position >= text.length) {
            controller.close();
            resolve();
            return;
          }

          const chunk = text.slice(position, position + chunkSize);
          position += chunkSize;

          controller.enqueue(chunk);

          resolve();
        }, delay);
      });
    },
  });
}
function clickHandle() {
  const stream = createStream(md);
  const containerDom = container.value!;
  createMarkdownStreamRender(stream, containerDom);
}
</script>
<template>
  <div class="w-screen h-screen py-16 mx-auto max-w-[800px]">
    <button
      class="bg-[#0a0a0a] text-white py-2 px-4 rounded-sm border border-[#262626] hover:bg-[#262626] cursor-pointer"
      @click="clickHandle"
    >
      Re-render
    </button>
    <div
      class="my-10 prose mt-10 dark:prose-invert"
      ref="container"
      style="max-width: 100%"
    ></div>
  </div>
</template>
<
