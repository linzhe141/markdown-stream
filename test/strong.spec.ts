import { getMarkdownStreamAst } from "../src";
import { expect, test } from "vitest";
import { createStream } from "./testUtils";

test(`basic  strong`, async () => {
  const md = `
**xxxx**

**\`linzhe\`**
`;
  const ast = await getMarkdownStreamAst(createStream(md));
  expect(ast.map((i) => JSON.stringify(i))).toMatchSnapshot();
});

test(`strong with inline code`, async () => {
  const md = `aaa**\`bar\` foo \`zzz\`**bbb`;
  const ast = await getMarkdownStreamAst(createStream(md));
  expect(ast.map((i) => JSON.stringify(i))).toMatchSnapshot();
});


test(`strong with one newline`, async () => {
  const md = `*foo\`bar\`zzzzzzzz*ggg
*xxxx* 
*\`linzhe\`*`;
  const ast = await getMarkdownStreamAst(createStream(md));
  expect(ast.map((i) => JSON.stringify(i))).toMatchSnapshot();
});
