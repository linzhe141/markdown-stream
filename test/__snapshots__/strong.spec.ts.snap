// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic  strong 1`] = `
[
  "{"operate":"blockOpen","type":"strongOpen","chunk":"**"}",
  "{"operate":"append","type":"strong","chunk":"xxxx"}",
  "{"operate":"append","type":"strongClose","chunk":"**"}",
  "{"operate":"blockOpen","type":"strongOpen","chunk":"**"}",
  "{"operate":"append","type":"inlineCodeOpen","chunk":"\`"}",
  "{"operate":"append","type":"inlineCode","chunk":"linzhe"}",
  "{"operate":"append","type":"inlineCodeClose","chunk":"\`"}",
  "{"operate":"append","type":"strongClose","chunk":"**"}",
]
`;

exports[`strong with inline code 1`] = `
[
  "{"operate":"blockOpen","type":"paragraph","chunk":"aaa"}",
  "{"operate":"append","type":"strongOpen","chunk":"**"}",
  "{"operate":"append","type":"inlineCodeOpen","chunk":"\`"}",
  "{"operate":"append","type":"inlineCode","chunk":"bar"}",
  "{"operate":"append","type":"inlineCodeClose","chunk":"\`"}",
  "{"operate":"append","type":"strong","chunk":" foo "}",
  "{"operate":"append","type":"inlineCodeOpen","chunk":"\`"}",
  "{"operate":"append","type":"inlineCode","chunk":"zzz"}",
  "{"operate":"append","type":"inlineCodeClose","chunk":"\`"}",
  "{"operate":"append","type":"strongClose","chunk":"**"}",
  "{"operate":"append","type":"paragraph","chunk":"bbb"}",
]
`;

exports[`strong with one newline 1`] = `
[
  "{"operate":"blockOpen","type":"italicOpen","chunk":"*"}",
  "{"operate":"append","type":"italic","chunk":"foo"}",
  "{"operate":"append","type":"inlineCodeOpen","chunk":"\`"}",
  "{"operate":"append","type":"inlineCode","chunk":"bar"}",
  "{"operate":"append","type":"inlineCodeClose","chunk":"\`"}",
  "{"operate":"append","type":"italic","chunk":"z"}",
  "{"operate":"append","type":"italic","chunk":"zzzzzzz"}",
  "{"operate":"append","type":"italicClose","chunk":"*"}",
  "{"operate":"append","type":"paragraph","chunk":"gg"}",
  "{"operate":"append","type":"paragraph","chunk":"g"}",
  "{"operate":"append","type":"italicOpen","chunk":"*"}",
  "{"operate":"append","type":"italic","chunk":"xxxx"}",
  "{"operate":"append","type":"italicClose","chunk":"*"}",
  "{"operate":"append","type":"paragraph","chunk":" "}",
  "{"operate":"append","type":"italicOpen","chunk":"*"}",
  "{"operate":"append","type":"inlineCodeOpen","chunk":"\`"}",
  "{"operate":"append","type":"inlineCode","chunk":"linzhe"}",
  "{"operate":"append","type":"inlineCodeClose","chunk":"\`"}",
  "{"operate":"append","type":"italicClose","chunk":"*"}",
]
`;
